### DISCUSSION TITLE
Text to speech aka Screenreader support

### DISCUSSION THREAD

**Discussion by Schr3da** [helix-editor/helix#10222] [Ideas]
Dear All,

Im a big fan of the project and use it as my vim replacement already - its a good editor and my daily tool. However it is not perfect but Im believing in this project as it did constantly great improvements. 

I would like to open the discussion about screen reader and accessibility. As I notice a lot of tools are not really blind people friendly - I would like to open the discussion here - to get some review. 

The basic idea is to add text to speech support of text words or lines under the cursor as well as menus. 
What do you think.

**Comment by archseer** [comment]
I'm not sure how to do this in a terminal application. Is there any protocol we can support?

**Comment by aleksanderkrauze** [comment]
Helix currently only supports terminal, but there seems to be motion to support multiple "frontends". So maybe we could tie screen-reader functionality to helix's core and only enable it on frontends that would support it.

**Comment by archseer** [comment]
`helix-core` is just a bunch of text editing primitives. Screen functionality is often frontend specific: I'm pretty sure you can't use the same API implementation on a GTK port that you would use on a mac app :)

**Comment by Schr3da** [comment]
I really would like to make it work in terminals as well. As different terminals use different render frameworks e.g. kitty is using opengl. So we could implement a service which gets shipped as part of helix? 

This would be generic and would not lock into some Window/framework specific thing. So if the service is enabled in the helix config the cursor (selected word, line ....) would than send the string to the module. Next the module converts it from text to speech. This could be done in multiple ways. What do you think?

**Comment by kirawi** [comment]
There is https://github.com/AccessKit/accesskit

**Comment by sempruijs** [comment]
I think a screen reader would be great! When I was blind, I could not find any editor with vim key bindings that was also accessible. vscode is in my opinion the most accessible editor for the visual impaired, and I did not find any plugins that enables vim keybindings that read the actions with VoiceOver (a screen reader for Mac) in the command line neither, not even Apple's own Xcode vim option was accessible. I was really sad at the time when I was depended on text-to-speach technology, because er was not a simple way how I could use vim. I think that blind developers would be really glad if helix becomes accessible for the blind. In the command line or a gui, just something so blind people can edit text fast. 🏃‍♀️👨‍🦯💻

**Comment by sempruijs** [comment]
So, I found [neoreader](https://github.com/MaxwellBo/neoreader). It is a screenreader plugin for neovim. I haven't tried it yet, but if you're searching for a accessible vim experience, I would recommend searching in that direction until helix has a screenreader.

**Comment by sempruijs** [comment]
Can we make a proof of concept with espeak? I found the following crate: https://docs.rs/espeaker/latest/espeaker/
No fancy accessible menu's or sound effects. Just when you type ```j``` it reads the next line and ```k``` for the previous line.
Just to see if it has potential. I myself am visually impaired and am willing to test. 
My dream is to have something like [emacspeak](https://github.com/tvraman/emacspeak) for helix.

**Comment by sempruijs** [comment]
Any progress on ideas for this?
